src/test/integration/IntegrationBase.t.sol

```diff
@@ -243,9 +286,91 @@ abstract contract IntegrationBase is IntegrationDeployer {
+    // since the last snapshot
+    function assert_Snap_Removed_TokenBalances(
+        User staker,
+        IStrategy[] memory strategies,
+        uint[] memory removedTokens,
+        string memory err
+    ) internal {
+        IERC20[] memory tokens = _getUnderlyingTokens(strategies);
+
+        uint[] memory curTokenBalances = _getTokenBalances(staker, tokens);
+        // Use timewarp to get previous token balances
+        uint[] memory prevTokenBalances = _getPrevTokenBalances(staker, tokens);
+
+        for (uint i = 0; i < tokens.length; i++) {
+            uint prevBalance = prevTokenBalances[i];
+            uint curBalance = curTokenBalances[i];
+
+            assertEq(prevBalance - removedTokens[i], curBalance, err);
+        }
+    }
+
+    /// @dev Check that the staker's underlying token balance for ALL tokens have
+    /// not changed since the last snapshot
+    function assert_Snap_Unchanged_TokenBalances(
+        User staker,
+        string memory err
+    ) internal {
+        IERC20[] memory tokens = allTokens;
+
+        uint[] memory curTokenBalances = _getTokenBalances(staker, tokens);
+        // Use timewarp to get previous token balances
+        uint[] memory prevTokenBalances = _getPrevTokenBalances(staker, tokens);
+
+        for (uint i = 0; i < tokens.length; i++) {
+            assertEq(prevTokenBalances[i], curTokenBalances[i], err);
+        }
+    }
+
+    /// Other snapshot assertions:
+
+    function assert_Snap_Added_QueuedWithdrawals(
+        User staker, 
+        IDelegationManager.Withdrawal[] memory withdrawals,
+        string memory err
+    ) internal {
+        uint curQueuedWithdrawals = _getCumulativeWithdrawals(staker);
+        // Use timewarp to get previous cumulative withdrawals
+        uint prevQueuedWithdrawals = _getPrevCumulativeWithdrawals(staker);
+
+        assertEq(prevQueuedWithdrawals + withdrawals.length, curQueuedWithdrawals, err);
+    }
+
+    /*******************************************************************************
+                                UTILITY METHODS
+    *******************************************************************************/
+
+    function _randPartialWithdrawals(
+        IStrategy[] memory strategies, 
+        uint[] memory shares
+    ) internal returns (IStrategy[] memory, uint[] memory) {
+        uint stratsToWithdraw = _randUint({ min: 1, max: strategies.length });
+
+        IStrategy[] memory withdrawStrats = new IStrategy[](stratsToWithdraw);
+        uint[] memory withdrawShares = new uint[](stratsToWithdraw);
+
+        for (uint i = 0; i < stratsToWithdraw; i++) {
+            // Random portion of shares to withdraw:
+            // 0: withdraw all
+            // 1: withdraw half
+            // we could get more granular, but we'd need to handle:
+            // - native eth shares can only be withdrawn in units of 1 gwei
+            // - LST shares have an exchange rate when ultimately converted to tokens
```

Comment:
To potentially be more expressive here
1. For Native ETH - rand int between 1 Gwei and actual amount staked. Multiply by 1 gwei. If > actual amount staked, then withdraw all.
2. For LST shares - rand int between (strategy.underlyingToSharesView(1 unit of token), total shares). Though not perfect, still gets us more optionality?

Reply:
Good notes. I updated it so that:
- native eth withdrawn is a random amount of gwei (at least 1), up to the user's share amount
- LSTs are just a random amount of shares (min 1), up to the user's share amount

Was surprisingly easy. I want to say I was preempting some future change that will complicate this a bit, but I can't remember what it was and this works great for now.

Reply:
[006e5b9](https://github.com/Layr-Labs/eigenlayer-contracts/pull/349/commits/006e5b901609b1e28390d3796e26859fab16e03a)

Reply:
i know what this was.

i figured we'd add balance updates and actual beacon chain partial withdrawals later, and wasn't sure how to work this in.

we'll cross that bridge when we come to it though!